oeps/oep-0038-Data-Modeling.rst

```diff
@@ -0,0 +1,418 @@
+
+   -  Updated date should be named: “modified”
+
+Foreign keys
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+-  Data should be joined using primary keys wherever possible
+
+-  Foreign keys should use a naming convention of :code:`<object_name>_id` where object name is the name of the table of the foreign key relationship.
+
+-  Do not join on things such as username, email address, or other dimensions of data that may change over time
+
+-  Do not join on PII
+
+- Joining between IDAs should be done by using a universally unique identifier (UUID)
+
+- In Django use `Attributes for fields with relations`_. to identify and link models with relationships.
+.. __Attributes for fields with relations: https://docs.djangoproject.com/en/2.2/ref/models/fields/#module-django.db.models.fields.related
+
+History for models involved with enrollments, courses and course metadata, or data involving financial payments and transactions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+-  History using django-simple-history_. .. _django-simple-history: https://django-simple-history.readthedocs.io/en/latest/
+
+-  Remember to `backfill history`_. for existing models.
+
+.. _backfill history: https://django-simple-history.readthedocs.io/en/latest/quick_start.html#existing-projects
+
+-  Where Django simple history is not an option, the following data should be captured:
+
+   -  Fields that were changed
+
+   -  Date & time of the change
+
+   -  The foreign key of of the user who initiated the change
+
+Use the correct data type for a column
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+-  Don’t use a :code:`IntegerField` when a :code:`BooleanField` would do.
+
+-  Use :code:`BigIntegerField`_. for your foreign keys
+
+-  Don’t store an Integer field as :code:`CharField`_..
+.. _CharField: https://docs.djangoproject.com/en/2.2/ref/models/fields/#charfield
+
+- Store UUID's as :code:`UUIDField`_. with a max length that matches the max length of the UUID.
+.. _UUIDField: https://docs.djangoproject.com/en/2.2/ref/models/fields/#uuidfield
+
+
+Each column in a table should only store a single fact or dimension
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- If a column could be a mix of integer data and character data it is best to store these items as 2 different fields in the database
+
+Models should be Annotated for PII
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+-  All models in the Open edX ecosystem should be tagged for PII using `code annotations`_. by following OEP-30_.
+.. _code annotations: https://github.com/edx/code-annotations>
+.. _OEP-30: https://github.com/edx/open-edx-proposals/blob/master/oeps/oep-0030-arch-pii-markup-and-auditing.rst
+
+
+Have sane default values
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+- A model should have default values that make sense for the application
+
+- :code:`CharField`_. Should be defined with the :code:`null=True`_. option.
```

Comment:
@macdiesel Django says the opposite, can you include the rationale here?

Reply:
The reasoning on this is that there is a difference between an empty string value and a null value. When we set the null=True option we know that no one has attempted to fill out that field vs someone had filled out and blanked the field. It's helpful for analysis.